Syntax extension API: allow to create extension nodes

nojb opened the pull request:
Currently it is not possible to create nodes defined by syntax extensions using the C API if they use the `opaque` field (e.g. the `"table"` extension) because there is no way to correctly allocate this field without exposing the details of the syntax extension in question. Concretely this means that it is not possible to programmatically create documents containing syntax extension nodes.

This PR adds a field `opaque_alloc_func`to the `cmark_syntax_extension` struct of type
```c
typedef void (*cmark_opaque_alloc_func) (cmark_syntax_extension *extension,
                                         cmark_mem *mem,
                                         cmark_node *node);
```
which is supposed to fill the `opaque` field of a `cmark_node` in those syntax extensions that need it.

Since this field must be allocated when creating the `cmark_node`, the function `cmark_node_new` has been extended to take a `cmark_syntax_extension *` argument:
```c
CMARK_EXPORT cmark_node *cmark_node_new(cmark_node_type type,
                                        cmark_syntax_extension *extension);
```
(a new function could be introduced for this purpose to avoid changing the signature of `cmark_node_new`.)

What do you think? Do you find this approach reasonable? Opinions welcome!

Thanks!

kivikakk commented:
> This PR adds a field `opaque_alloc_func`to the `cmark_syntax_extension`

I think this is great!

> (a new function could be introduced for this purpose to avoid changing the signature of `cmark_node_new`.)

From an API design perspective, I think this would be best.

Thank you so much for the PR! I'd be happy to merge with this last change, and whatever else is needed to get the tests passing again. (Looks like it's the `cmark_node_new` signature change, and maybe some new warnings re: redefining `cmark_syntax_extension`?)

nojb commented:
Sorry, I was quite busy for the last couple of months and could not see to this, but I have now cleaned up the patch, and introduced a separate function as per discussion above. All tests pass. Let me know if you have any other observations. Thanks!

kivikakk commented:
Thank you so much! It's looking great 👍

nojb commented:
Thanks!
